typescript/sdk/src/core/HyperlaneCore.ts

```diff
@@ -175,19 +165,129 @@ export class HyperlaneCore extends HyperlaneApp<CoreFactories> {
     };
   }
 
-  async relayMessage(
+  onDispatch(
+    handler: (
+      message: DispatchedMessage,
+      event: DispatchEvent,
+    ) => Promise<void>,
+    chains = Object.keys(this.contractsMap),
+  ): {
+    removeHandler: (chains?: ChainName[]) => void;
+  } {
+    chains.map((originChain) => {
+      const mailbox = this.contractsMap[originChain].mailbox;
+      this.logger.debug(`Listening for dispatch on ${originChain}`);
+      mailbox.on<DispatchEvent>(
+        mailbox.filters.Dispatch(),
+        (_sender, _destination, _recipient, message, event) => {
+          const parsed = HyperlaneCore.parseDispatchedMessage(message);
+          this.logger.info(`Observed message ${parsed.id} on ${originChain}`);
+          return handler(parsed, event);
+        },
+      );
+    });
+
+    return {
+      removeHandler: (removeChains) =>
+        (removeChains ?? chains).map((originChain) => {
+          this.contractsMap[originChain].mailbox.removeAllListeners('Dispatch');
+          this.logger.debug(`Stopped listening for dispatch on ${originChain}`);
+        }),
+    };
+  }
+
+  getDefaults(): Promise<ChainMap<{ ism: Address; hook: Address }>> {
+    return promiseObjAll(
+      objMap(this.contractsMap, async (_, contracts) => ({
+        ism: await contracts.mailbox.defaultIsm(),
+        hook: await contracts.mailbox.defaultHook(),
+      })),
+    );
+  }
+
+  getIsm(
+    destinationChain: ChainName,
+    recipientAddress: Address,
+  ): Promise<Address> {
+    const destinationMailbox = this.contractsMap[destinationChain];
+    return destinationMailbox.mailbox.recipientIsm(recipientAddress);
+  }
+
+  protected getRecipient(message: DispatchedMessage): IMessageRecipient {
+    return IMessageRecipient__factory.connect(
+      bytes32ToAddress(message.parsed.recipient),
+      this.multiProvider.getProvider(this.getDestination(message)),
+    );
+  }
+
+  estimateHandle(message: DispatchedMessage): Promise<ethers.BigNumber> {
```

Comment:
Recommend against using ethers BN types in our class interfaces. It's too evm specific, and specific too an outdated lib. `string` would be fine. `bigint` would be better but too inconsistent with what we've done elsewhere

Reply:
switched to string